docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,206 @@
+
+
+### Migration workflow
+1. Starting migration
+    - Migration can only be started after a Shoot cluster has been successfully created so that the `status.seed` field in the `Shoot` resource has been set
+    - The `Shoot` resource's field `spec.seedName="new-seed"` is edited to hold the name of the _New Seed_ and reconciliation is automatically triggered
+    - The Garden Controller Manager checks if the equality between `spec.seedName` and `status.seed`, detects that they are different and triggers migration.
+1. The Garden Controller Manager waits for the _New Seed_ to be ready
+1. ShootState is synced from the _Host Seed_ cluster
+    - The Garden Controller Manager lists all deployed extension resources for the Shoot and copies the values in their `status.state` field to the `ShootState` resource
+    - If the _Host Seed's_ Apiserver is not available, then last successfull synced `ShotoState` is used
+1. Stop the Shoot's api server
+1. Scale Down the Shoot's control plane in the _Host Seed_. ETCD backup will be done automatically before the ETCD pod is scaled down.
+1. Resume normal reconciliation flow in the _New Seed_. Extension resources are annotated with `core.gardener.cloud/operation=restore` to instruct the extension controllers to reconstruct their state.
+
+
+### Leader Election and Control Plane Termination
+
+During migration "split brain" scenario must be avoided. This means that a Shoot's control plane in the _Host Seed_ must be scaled down before it is scaled up in the _New Seed_.
+
+**Note**: This section is still under discussion. The plan is to first implement the `ShootState` and modify the reconcilation flow and extensions accordingly. Additionally multiple scenarios need to be considered depending on the reachability to the Garden cluster from the Host Seed components, the Host Seed's api server from the Garden cluster and the Host Seed's api server from the controllers running on the seed. The initial implementation will only cover the case where everything is running.
+
+Extension controllers do not need leader election functionality because they only reconcile the extension resources if the reconcile operation annotation is specified in the resource. Since this is set only during a reconciliation triggered by the Garden Controller Manager it cannot happen during migration.
+
+For other controllers in the controlplane (e.g MCM, etcd backup restore, kube apiserver) leader election has to be implemented. We plan to introduce gardenlet soon therefore the garden cluster is expected to be reachable from all seeds. Initially the Shoot's `spec.seedName` can be used to determine which seed is the current leader of the shoot.
```

Comment:
Yeah, well, unless there are network issues and you don't want a healthy seed with hundreds of shoots to shut down, if it's separated from the Garden cluster. That will be tricky. DNS is bad/complex due to TTLs, but at least it's 100% available (AWS or GCP both guarantee 100% availability, but also they get attacked).

This will be a hard decision to take. The only safe means is revocation of access, at least that was the outcome of this discussion last time.

Reply:
We couldn't find a good way to handle the scenario when the seed components cannot talk to w/e component is providing the Leader Election. DNS can be complex due to both the necessary setup in the Garden cluster and TTLs. Also there are still some DNS servers which do not respect TTLs.